src/3rdparty/squirrel/squirrel/sqobject.h

```diff
@@ -344,13 +345,48 @@ struct SQCollectable : public SQRefCounted {
 	SQCollectable *_prev;
 	SQSharedState *_sharedstate;
 	virtual void Release()=0;
-	virtual void Mark(SQCollectable **chain)=0;
+	virtual void EnqueueMarkObjectForChildren(class SQGCMarkerQueue &queue)=0;
 	void UnMark();
 	virtual void Finalize()=0;
 	static void AddToChain(SQCollectable **chain,SQCollectable *c);
 	static void RemoveFromChain(SQCollectable **chain,SQCollectable *c);
 };
 
+/**
+ * Helper container for state to change the garbage collection from a recursive to an iterative approach.
+ * The iterative approach provides effectively a depth first search approach.
+ */
+class SQGCMarkerQueue {
+	std::forward_list<SQCollectable*> queue; ///< The queue of elements to still process.
```

Comment:
I'd suggest benchmarking using a std::vector instead here. A linked list might have poor memory locality performance with lots of tiny allocations like this.

Reply:
Implementing the same logic as with the forward list decreases the performance of the GC marking by a factor of about ~2. Rethinking the logic to better suit the vector, i.e. the most recent elements are at the end instead of the begin, improves the performance by about a factor of ~2. This all in a debug build, but I would reckon the same effect holds for the release builds.

Going from a deque to a vector for the "free" queue is more cumbersome as it now requires some extra state information whether processing is happening or not (to prevent removing an element somewhere in the middle of the vector), though there the factor of ~2 improvement could be observed as well. So this suggestion definitely helped.

Testing was done with a custom AI that created a linked list of 500 000 elements, and a stacked array and table of depth 500 000, so 1.5 million objects in total. Then the average of 5 GC runs was taken (begin of CollectGarbage until the queue was empty), and for the Release just from setting _collectable_free_processing to true until setting it to false. All with a debug build on not sure powerful machine. 

GC marking:
* Forward list: 1676 ms
* Vector insert front, remove front: 3257 ms (~194% of original)
* Vector push back, pop back: 688 ms (~41% of original)

Free:
* Forward list: 2577
* Vector push back, pop back: 1466 (~57% of original)